// @n8n/n8n-nodes-langchain/dist/nodes/embeddings/EmbeddingsLemonade/EmbeddingsLemonade.node.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var EmbeddingsLemonade_node_exports = {};
__export(EmbeddingsLemonade_node_exports, {
  EmbeddingsLemonade: () => EmbeddingsLemonade
});
module.exports = __toCommonJS(EmbeddingsLemonade_node_exports);
var import_openai = require("@langchain/openai");
var import_n8n_workflow = require("n8n-workflow");
var import_logWrapper = require("../../../utils/logWrapper");
var import_sharedFields = require("../../../utils/sharedFields");
var import_description = require("../../llms/LMLemonade/description");
class EmbeddingsLemonade {
  constructor() {
    this.description = {
      displayName: "Embeddings Lemonade",
      name: "embeddingsLemonade",
      icon: "file:lemonade.svg",
      group: ["transform"],
      version: 1,
      description: "Use Lemonade Embeddings",
      defaults: {
        name: "Embeddings Lemonade"
      },
      ...import_description.lemonadeDescription,
      codex: {
        categories: ["AI"],
        subcategories: {
          AI: ["Embeddings"]
        },
        resources: {
          primaryDocumentation: [
            {
              url: "https://docs.n8n.io/integrations/builtin/cluster-nodes/sub-nodes/n8n-nodes-langchain.embeddingslemonade/"
            }
          ]
        }
      },
      inputs: [],
      outputs: [import_n8n_workflow.NodeConnectionTypes.AiEmbedding],
      outputNames: ["Embeddings"],
      properties: [(0, import_sharedFields.getConnectionHintNoticeField)([import_n8n_workflow.NodeConnectionTypes.AiVectorStore]), import_description.lemonadeModel]
    };
  }
  async supplyData(itemIndex) {
    const modelName = this.getNodeParameter("model", itemIndex);
    const credentials = await this.getCredentials("lemonadeApi");
    const apiKey = credentials.apiKey || "lemonade-placeholder-key";
    const configuration = {
      baseURL: credentials.baseUrl
    };
    if (credentials.apiKey) {
      configuration.defaultHeaders = {
        Authorization: `Bearer ${credentials.apiKey}`
      };
    }
    const embeddings = new import_openai.OpenAIEmbeddings({
      apiKey,
      model: modelName,
      configuration
    });
    return {
      response: (0, import_logWrapper.logWrapper)(embeddings, this)
    };
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  EmbeddingsLemonade
});
