// @n8n/n8n-nodes-langchain/dist/credentials/LemonadeApi.credentials.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var LemonadeApi_credentials_exports = {};
__export(LemonadeApi_credentials_exports, {
  LemonadeApi: () => LemonadeApi
});
module.exports = __toCommonJS(LemonadeApi_credentials_exports);
class LemonadeApi {
  constructor() {
    this.name = "lemonadeApi";
    this.displayName = "Lemonade";
    this.documentationUrl = "lemonade";
    this.properties = [
      {
        displayName: "Base URL",
        name: "baseUrl",
        required: true,
        type: "string",
        default: "http://localhost:8000/api/v1"
      },
      {
        displayName: "API Key",
        hint: "Optional API key for Lemonade server authentication. Not required for default Lemonade installation",
        name: "apiKey",
        type: "string",
        typeOptions: { password: true },
        default: "",
        required: false
      }
    ];
    this.test = {
      request: {
        baseURL: "={{ $credentials.baseUrl }}",
        url: "/models",
        method: "GET"
      }
    };
  }
  async authenticate(credentials, requestOptions) {
    const apiKey = credentials.apiKey;
    if (apiKey && apiKey.trim() !== "") {
      requestOptions.headers = {
        ...requestOptions.headers,
        Authorization: `Bearer ${apiKey}`
      };
    }
    return requestOptions;
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  LemonadeApi
});
