// @n8n/n8n-nodes-langchain/dist/credentials/McpOAuth2Api.credentials.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var McpOAuth2Api_credentials_exports = {};
__export(McpOAuth2Api_credentials_exports, {
  McpOAuth2Api: () => McpOAuth2Api
});
module.exports = __toCommonJS(McpOAuth2Api_credentials_exports);
class McpOAuth2Api {
  constructor() {
    this.name = "mcpOAuth2Api";
    this.extends = ["oAuth2Api"];
    this.displayName = "MCP OAuth2 API";
    this.documentationUrl = "mcp";
    this.properties = [
      {
        displayName: "Use Dynamic Client Registration",
        name: "useDynamicClientRegistration",
        type: "boolean",
        default: true,
        required: true
      }
    ];
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  McpOAuth2Api
});
