// @n8n/n8n-nodes-langchain/dist/nodes/llms/LMLemonade/description.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var description_exports = {};
__export(description_exports, {
  lemonadeDescription: () => lemonadeDescription,
  lemonadeModel: () => lemonadeModel,
  lemonadeOptions: () => lemonadeOptions
});
module.exports = __toCommonJS(description_exports);
const lemonadeDescription = {
  credentials: [
    {
      name: "lemonadeApi",
      required: true
    }
  ],
  requestDefaults: {
    ignoreHttpStatusErrors: true,
    baseURL: '={{ $credentials.baseUrl.replace(new RegExp("/$"), "") }}'
  }
};
const lemonadeModel = {
  displayName: "Model",
  name: "model",
  type: "options",
  default: "",
  description: "The model which will generate the completion. Models are loaded and managed through the Lemonade server.",
  typeOptions: {
    loadOptions: {
      routing: {
        request: {
          method: "GET",
          url: "/models"
        },
        output: {
          postReceive: [
            {
              type: "rootProperty",
              properties: {
                property: "data"
              }
            },
            {
              type: "setKeyValue",
              properties: {
                name: "={{$responseItem.id}}",
                value: "={{$responseItem.id}}"
              }
            },
            {
              type: "sort",
              properties: {
                key: "name"
              }
            }
          ]
        }
      }
    }
  },
  routing: {
    send: {
      type: "body",
      property: "model"
    }
  },
  required: true
};
const lemonadeOptions = {
  displayName: "Options",
  name: "options",
  placeholder: "Add Option",
  description: "Additional options to add",
  type: "collection",
  default: {},
  options: [
    {
      displayName: "Sampling Temperature",
      name: "temperature",
      default: 0.7,
      typeOptions: { maxValue: 2, minValue: 0, numberPrecision: 1 },
      description: "Controls the randomness of the generated text. Lower values make the output more focused and deterministic, while higher values make it more diverse and random.",
      type: "number"
    },
    {
      displayName: "Top P",
      name: "topP",
      default: 1,
      typeOptions: { maxValue: 1, minValue: 0, numberPrecision: 1 },
      description: "Chooses from the smallest possible set of tokens whose cumulative probability exceeds the probability top_p. Helps generate more human-like text by reducing repetitions.",
      type: "number"
    },
    {
      displayName: "Frequency Penalty",
      name: "frequencyPenalty",
      type: "number",
      default: 0,
      typeOptions: { minValue: -2, maxValue: 2, numberPrecision: 1 },
      description: "Adjusts the penalty for tokens that have already appeared in the generated text. Positive values discourage repetition, negative values encourage it."
    },
    {
      displayName: "Presence Penalty",
      name: "presencePenalty",
      type: "number",
      default: 0,
      typeOptions: { minValue: -2, maxValue: 2, numberPrecision: 1 },
      description: "Adjusts the penalty for tokens based on their presence in the generated text so far. Positive values penalize tokens that have already appeared, encouraging diversity."
    },
    {
      displayName: "Max Tokens to Generate",
      name: "maxTokens",
      type: "number",
      default: -1,
      description: "The maximum number of tokens to generate. Set to -1 for no limit. Be cautious when setting this to a large value, as it can lead to very long outputs."
    },
    {
      displayName: "Stop Sequences",
      name: "stop",
      type: "string",
      default: "",
      description: "Comma-separated list of sequences where the model will stop generating text"
    }
  ]
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  lemonadeDescription,
  lemonadeModel,
  lemonadeOptions
});
