// @n8n/n8n-nodes-langchain/dist/nodes/vector_store/shared/createVectorStoreNode/operations/retrieveAsToolExecuteOperation.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var retrieveAsToolExecuteOperation_exports = {};
__export(retrieveAsToolExecuteOperation_exports, {
  handleRetrieveAsToolExecuteOperation: () => handleRetrieveAsToolExecuteOperation
});
module.exports = __toCommonJS(retrieveAsToolExecuteOperation_exports);
var import_n8n_workflow = require("n8n-workflow");
var import_helpers = require("../../../../../utils/helpers");
async function handleRetrieveAsToolExecuteOperation(context, args, embeddings, itemIndex) {
  const filter = (0, import_helpers.getMetadataFiltersValues)(context, itemIndex);
  const vectorStore = await args.getVectorStoreClient(
    context,
    // We'll pass filter to similaritySearchVectorWithScore instead of getVectorStoreClient
    void 0,
    embeddings,
    itemIndex
  );
  try {
    const inputData = context.getInputData();
    const item = inputData[itemIndex];
    const query = typeof item.json.input === "string" ? item.json.input : void 0;
    if (!query || typeof query !== "string") {
      throw new Error('Input data must contain a "input" field with the search query');
    }
    const topK = context.getNodeParameter("topK", itemIndex, 4);
    (0, import_n8n_workflow.assertParamIsNumber)("topK", topK, context.getNode());
    const useReranker = context.getNodeParameter("useReranker", itemIndex, false);
    (0, import_n8n_workflow.assertParamIsBoolean)("useReranker", useReranker, context.getNode());
    const includeDocumentMetadata = context.getNodeParameter(
      "includeDocumentMetadata",
      itemIndex,
      true
    );
    (0, import_n8n_workflow.assertParamIsBoolean)("includeDocumentMetadata", includeDocumentMetadata, context.getNode());
    const embeddedQuery = await embeddings.embedQuery(query);
    let docs = await vectorStore.similaritySearchVectorWithScore(embeddedQuery, topK, filter);
    if (useReranker && docs.length > 0) {
      const reranker = await context.getInputConnectionData(
        import_n8n_workflow.NodeConnectionTypes.AiReranker,
        0
      );
      const documents = docs.map(([doc]) => doc);
      const rerankedDocuments = await reranker.compressDocuments(documents, query);
      docs = rerankedDocuments.map((doc) => {
        const { relevanceScore, ...metadata } = doc.metadata || {};
        return [{ ...doc, metadata }, relevanceScore ?? 0];
      });
    }
    const serializedDocs = docs.map(([doc]) => {
      if (includeDocumentMetadata) {
        return {
          type: "text",
          text: JSON.stringify({ ...doc })
        };
      } else {
        return {
          type: "text",
          pageContent: JSON.stringify({ pageContent: doc.pageContent })
        };
      }
    });
    (0, import_helpers.logAiEvent)(context, "ai-vector-store-searched", { input: query });
    return [
      {
        json: {
          response: serializedDocs
        },
        pairedItem: {
          item: itemIndex
        }
      }
    ];
  } finally {
    args.releaseVectorStoreClient?.(vectorStore);
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  handleRetrieveAsToolExecuteOperation
});
