// @n8n/n8n-nodes-langchain/dist/nodes/vendors/OpenAi/v1/OpenAiV1.node.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var OpenAiV1_node_exports = {};
__export(OpenAiV1_node_exports, {
  OpenAiV1: () => OpenAiV1
});
module.exports = __toCommonJS(OpenAiV1_node_exports);
var import_n8n_workflow = require("n8n-workflow");
var import_methods = require("../methods");
var import_router = require("./actions/router");
var import_description = require("../helpers/description");
var assistant = __toESM(require("./actions/assistant"));
var audio = __toESM(require("./actions/audio"));
var file = __toESM(require("./actions/file"));
var image = __toESM(require("./actions/image"));
var text = __toESM(require("./actions/text"));
class OpenAiV1 {
  constructor(baseDescription) {
    this.methods = {
      listSearch: import_methods.listSearch,
      loadOptions: import_methods.loadOptions
    };
    this.description = {
      ...baseDescription,
      version: [1, 1.1, 1.2, 1.3, 1.4, 1.5, 1.6, 1.7, 1.8],
      defaults: {
        name: "OpenAI"
      },
      inputs: `={{(${import_description.configureNodeInputs})($parameter.resource, $parameter.operation, $parameter.hideTools, $parameter.memory ?? undefined)}}`,
      outputs: [import_n8n_workflow.NodeConnectionTypes.Main],
      credentials: [
        {
          name: "openAiApi",
          required: true
        }
      ],
      properties: [
        {
          displayName: "Resource",
          name: "resource",
          type: "options",
          noDataExpression: true,
          // eslint-disable-next-line n8n-nodes-base/node-param-options-type-unsorted-items
          options: [
            {
              name: "Assistant",
              value: "assistant"
            },
            {
              name: "Text",
              value: "text"
            },
            {
              name: "Image",
              value: "image"
            },
            {
              name: "Audio",
              value: "audio"
            },
            {
              name: "File",
              value: "file"
            }
          ],
          default: "text"
        },
        ...assistant.description,
        ...audio.description,
        ...file.description,
        ...image.description,
        ...text.description
      ]
    };
  }
  async execute() {
    return await import_router.router.call(this);
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  OpenAiV1
});
