// @n8n/n8n-nodes-langchain/dist/nodes/Guardrails/helpers/configureNodeInputs.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var configureNodeInputs_exports = {};
__export(configureNodeInputs_exports, {
  configureNodeInputs: () => configureNodeInputs
});
module.exports = __toCommonJS(configureNodeInputs_exports);
const configureNodeInputs = (operation) => {
  if (operation === "sanitize") {
    return ["main"];
  }
  return [
    "main",
    {
      type: "ai_languageModel",
      displayName: "Chat Model",
      maxConnections: 1,
      required: true,
      filter: {
        excludedNodes: [
          "@n8n/n8n-nodes-langchain.lmCohere",
          "@n8n/n8n-nodes-langchain.lmOllama",
          "n8n/n8n-nodes-langchain.lmOpenAi",
          "@n8n/n8n-nodes-langchain.lmOpenHuggingFaceInference"
        ]
      }
    }
  ];
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  configureNodeInputs
});
