// @n8n/n8n-nodes-langchain/dist/nodes/llms/LMChatLemonade/LmChatLemonade.node.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var LmChatLemonade_node_exports = {};
__export(LmChatLemonade_node_exports, {
  LmChatLemonade: () => LmChatLemonade
});
module.exports = __toCommonJS(LmChatLemonade_node_exports);
var import_openai = require("@langchain/openai");
var import_n8n_workflow = require("n8n-workflow");
var import_sharedFields = require("../../../utils/sharedFields");
var import_description = require("../LMLemonade/description");
var import_n8nLlmFailedAttemptHandler = require("../n8nLlmFailedAttemptHandler");
var import_N8nLlmTracing = require("../N8nLlmTracing");
class LmChatLemonade {
  constructor() {
    this.description = {
      displayName: "Lemonade Chat Model",
      name: "lmChatLemonade",
      icon: "file:lemonade.svg",
      group: ["transform"],
      version: 1,
      description: "Language Model Lemonade Chat",
      defaults: {
        name: "Lemonade Chat Model"
      },
      codex: {
        categories: ["AI"],
        subcategories: {
          AI: ["Language Models", "Root Nodes"],
          "Language Models": ["Chat Models (Recommended)"]
        },
        resources: {
          primaryDocumentation: [
            {
              url: "https://docs.n8n.io/integrations/builtin/cluster-nodes/sub-nodes/n8n-nodes-langchain.lmchatlemonade/"
            }
          ]
        }
      },
      inputs: [],
      outputs: [import_n8n_workflow.NodeConnectionTypes.AiLanguageModel],
      outputNames: ["Model"],
      ...import_description.lemonadeDescription,
      properties: [
        (0, import_sharedFields.getConnectionHintNoticeField)([import_n8n_workflow.NodeConnectionTypes.AiChain, import_n8n_workflow.NodeConnectionTypes.AiAgent]),
        import_description.lemonadeModel,
        import_description.lemonadeOptions
      ]
    };
  }
  async supplyData(itemIndex) {
    const credentials = await this.getCredentials("lemonadeApi");
    const modelName = this.getNodeParameter("model", itemIndex);
    const options = this.getNodeParameter("options", itemIndex, {});
    const processedOptions = {
      ...options,
      maxTokens: options.maxTokens && options.maxTokens > 0 ? options.maxTokens : void 0,
      stop: void 0
      // Will be set below if options.stop exists
    };
    if (options.stop) {
      const stopSequences = options.stop.split(",").map((s) => s.trim()).filter((s) => s.length > 0);
      processedOptions.stop = stopSequences.length > 0 ? stopSequences : void 0;
    }
    const configuration = {
      baseURL: credentials.baseUrl
    };
    if (credentials.apiKey) {
      configuration.defaultHeaders = {
        Authorization: `Bearer ${credentials.apiKey}`
      };
    }
    const model = new import_openai.ChatOpenAI({
      apiKey: credentials.apiKey || "lemonade-placeholder-key",
      model: modelName,
      ...processedOptions,
      configuration,
      callbacks: [new import_N8nLlmTracing.N8nLlmTracing(this)],
      onFailedAttempt: (0, import_n8nLlmFailedAttemptHandler.makeN8nLlmFailedAttemptHandler)(this)
    });
    return {
      response: model
    };
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  LmChatLemonade
});
