// n8n/dist/modules/mcp/mcp-server-middleware.service.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.McpServerMiddlewareService = void 0;
const di_1 = require("@n8n/di");
const auth_error_1 = require("../../errors/response-errors/auth.error");
const jwt_service_1 = require("../../services/jwt.service");
const telemetry_1 = require("../../telemetry");
const mcp_api_key_service_1 = require("./mcp-api-key.service");
const mcp_oauth_token_service_1 = require("./mcp-oauth-token.service");
const mcp_constants_1 = require("./mcp.constants");
const mcp_utils_1 = require("./mcp.utils");
let McpServerMiddlewareService = class McpServerMiddlewareService {
    constructor(mcpServerApiKeyService, mcpAuthTokenService, jwtService, telemetry) {
        this.mcpServerApiKeyService = mcpServerApiKeyService;
        this.mcpAuthTokenService = mcpAuthTokenService;
        this.jwtService = jwtService;
        this.telemetry = telemetry;
    }
    async getUserForToken(token) {
        let decoded;
        try {
            decoded = this.jwtService.decode(token);
        }
        catch (error) {
            return null;
        }
        if (decoded?.meta?.isOAuth === true) {
            return await this.mcpAuthTokenService.verifyOAuthAccessToken(token);
        }
        return await this.mcpServerApiKeyService.verifyApiKey(token);
    }
    getAuthMiddleware() {
        return async (req, res, next) => {
            const authorizationHeader = req.header('authorization');
            if (!authorizationHeader) {
                this.responseWithUnauthorized(res, req);
                return;
            }
            const token = this.extractBearerToken(authorizationHeader);
            if (!token) {
                this.responseWithUnauthorized(res, req);
                return;
            }
            const user = await this.getUserForToken(token);
            if (!user) {
                this.responseWithUnauthorized(res, req);
                return;
            }
            req.user = user;
            next();
        };
    }
    extractBearerToken(headerValue) {
        if (!headerValue.startsWith('Bearer')) {
            throw new auth_error_1.AuthError('Invalid authorization header format');
        }
        const tokenMatch = headerValue.match(/^Bearer\s+(.+)$/i);
        if (tokenMatch) {
            return tokenMatch[1];
        }
        throw new auth_error_1.AuthError('Invalid authorization header format');
    }
    responseWithUnauthorized(res, req) {
        this.trackUnauthorizedEvent(req);
        res.header('WWW-Authenticate', 'Bearer realm="n8n MCP Server"');
        res.status(401).send({ message: mcp_constants_1.UNAUTHORIZED_ERROR_MESSAGE });
    }
    trackUnauthorizedEvent(req) {
        const clientInfo = (0, mcp_utils_1.getClientInfo)(req);
        this.telemetry.track(mcp_constants_1.USER_CONNECTED_TO_MCP_EVENT, {
            mcp_connection_status: 'error',
            error: mcp_constants_1.UNAUTHORIZED_ERROR_MESSAGE,
            client_name: clientInfo?.name,
            client_version: clientInfo?.version,
        });
    }
};
exports.McpServerMiddlewareService = McpServerMiddlewareService;
exports.McpServerMiddlewareService = McpServerMiddlewareService = __decorate([
    (0, di_1.Service)(),
    __metadata("design:paramtypes", [mcp_api_key_service_1.McpServerApiKeyService,
        mcp_oauth_token_service_1.McpOAuthTokenService,
        jwt_service_1.JwtService,
        telemetry_1.Telemetry])
], McpServerMiddlewareService);
