// @n8n/api-types/dist/chat-hub.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ChatHubUpdateAgentRequest = exports.ChatHubCreateAgentRequest = exports.ChatHubUpdateConversationRequest = exports.ChatHubEditMessageRequest = exports.ChatHubRegenerateMessageRequest = exports.ChatHubSendMessageRequest = exports.emptyChatModelsResponse = exports.chatModelsRequestSchema = exports.chatHubConversationModelSchema = exports.PROVIDER_CREDENTIAL_TYPE_MAP = exports.chatHubProviderSchema = exports.chatHubLLMProviderSchema = void 0;
const zod_1 = require("zod");
const zod_class_1 = require("zod-class");
exports.chatHubLLMProviderSchema = zod_1.z.enum(['openai', 'anthropic', 'google']);
exports.chatHubProviderSchema = zod_1.z.enum([
    ...exports.chatHubLLMProviderSchema.options,
    'n8n',
    'custom-agent',
]);
exports.PROVIDER_CREDENTIAL_TYPE_MAP = {
    openai: 'openAiApi',
    anthropic: 'anthropicApi',
    google: 'googlePalmApi',
};
const openAIModelSchema = zod_1.z.object({
    provider: zod_1.z.literal('openai'),
    model: zod_1.z.string(),
});
const anthropicModelSchema = zod_1.z.object({
    provider: zod_1.z.literal('anthropic'),
    model: zod_1.z.string(),
});
const googleModelSchema = zod_1.z.object({
    provider: zod_1.z.literal('google'),
    model: zod_1.z.string(),
});
const n8nModelSchema = zod_1.z.object({
    provider: zod_1.z.literal('n8n'),
    workflowId: zod_1.z.string(),
});
const chatAgentSchema = zod_1.z.object({
    provider: zod_1.z.literal('custom-agent'),
    agentId: zod_1.z.string(),
});
exports.chatHubConversationModelSchema = zod_1.z.discriminatedUnion('provider', [
    openAIModelSchema,
    anthropicModelSchema,
    googleModelSchema,
    n8nModelSchema,
    chatAgentSchema,
]);
exports.chatModelsRequestSchema = zod_1.z.object({
    credentials: zod_1.z.record(exports.chatHubProviderSchema, zod_1.z.string().nullable()),
});
exports.emptyChatModelsResponse = {
    openai: { models: [] },
    anthropic: { models: [] },
    google: { models: [] },
    n8n: { models: [] },
    'custom-agent': { models: [] },
};
class ChatHubSendMessageRequest extends zod_class_1.Z.class({
    messageId: zod_1.z.string().uuid(),
    sessionId: zod_1.z.string().uuid(),
    message: zod_1.z.string(),
    model: exports.chatHubConversationModelSchema,
    previousMessageId: zod_1.z.string().uuid().nullable(),
    credentials: zod_1.z.record(zod_1.z.object({
        id: zod_1.z.string(),
        name: zod_1.z.string(),
    })),
}) {
}
exports.ChatHubSendMessageRequest = ChatHubSendMessageRequest;
class ChatHubRegenerateMessageRequest extends zod_class_1.Z.class({
    model: exports.chatHubConversationModelSchema,
    credentials: zod_1.z.record(zod_1.z.object({
        id: zod_1.z.string(),
        name: zod_1.z.string(),
    })),
}) {
}
exports.ChatHubRegenerateMessageRequest = ChatHubRegenerateMessageRequest;
class ChatHubEditMessageRequest extends zod_class_1.Z.class({
    message: zod_1.z.string(),
    messageId: zod_1.z.string().uuid(),
    model: exports.chatHubConversationModelSchema,
    credentials: zod_1.z.record(zod_1.z.object({
        id: zod_1.z.string(),
        name: zod_1.z.string(),
    })),
}) {
}
exports.ChatHubEditMessageRequest = ChatHubEditMessageRequest;
class ChatHubUpdateConversationRequest extends zod_class_1.Z.class({
    title: zod_1.z.string().optional(),
    credentialId: zod_1.z.string().max(36).optional(),
    provider: exports.chatHubProviderSchema.optional(),
    model: zod_1.z.string().max(64).optional(),
    workflowId: zod_1.z.string().max(36).optional(),
    agentId: zod_1.z.string().uuid().optional(),
}) {
}
exports.ChatHubUpdateConversationRequest = ChatHubUpdateConversationRequest;
class ChatHubCreateAgentRequest extends zod_class_1.Z.class({
    name: zod_1.z.string().min(1).max(128),
    description: zod_1.z.string().max(512).optional(),
    systemPrompt: zod_1.z.string().min(1),
    credentialId: zod_1.z.string(),
    provider: exports.chatHubProviderSchema.exclude(['n8n', 'custom-agent']),
    model: zod_1.z.string().max(64),
}) {
}
exports.ChatHubCreateAgentRequest = ChatHubCreateAgentRequest;
class ChatHubUpdateAgentRequest extends zod_class_1.Z.class({
    name: zod_1.z.string().min(1).max(128).optional(),
    description: zod_1.z.string().max(512).optional(),
    systemPrompt: zod_1.z.string().min(1).optional(),
    credentialId: zod_1.z.string().optional(),
    provider: exports.chatHubProviderSchema.optional(),
    model: zod_1.z.string().max(64).optional(),
}) {
}
exports.ChatHubUpdateAgentRequest = ChatHubUpdateAgentRequest;
