// n8n-editor-ui/dist/assets/NodeCreation-CL1i8pHd.js
const __vite__mapDeps=(i,m=__vite__mapDeps,d=(m.f||(m.f=["assets/NodeCreator-DEx3Dj6j.js","assets/_plugin-vue_export-helper-BwBpWJRZ.js","assets/src-DsaRShTK.js","assets/preload-helper-CR0ecmWK.js","assets/icon-DmIKo5zZ.js","assets/vue.runtime.esm-bundler-DDuXT-9r.js","assets/chunk-6z4oVpB-.js","assets/truncate-7XypoloM.js","assets/_MapCache-Dw6Llo22.js","assets/sanitize-html-Cfxibw1n.js","assets/empty-RTEMfF4N.js","assets/path-browserify-CTmc1OxV.js","assets/en-BYTsM8fR.js","assets/src-sI-DsSQQ.css","assets/nodeCreator.store-C1GkQ_2N.js","assets/constants-C5Tj3dXQ.js","assets/merge-Dnkh07HW.js","assets/useTelemetry-DedEakwK.js","assets/dateformat-D7TIhVd4.js","assets/useDebounce-DFa0sFlC.js","assets/useExternalHooks-DSSTcvme.js","assets/nodeIcon-ChI78z3o.js","assets/templates.store-Bt8tpLPx.js","assets/cloudPlan.store-DT2PFV79.js","assets/exports-D2D_Xd3F.js","assets/useToast-LQk5sJH0.js","assets/useCanvasOperations-Cxgmc_7x.js","assets/useClipboard-BkbYyMX3.js","assets/usePinnedData-BsuIBj2b.js","assets/focusPanel.store-CPU1uVq-.js","assets/folders.store-CmYx4Y_F.js","assets/executions.store-D2xBmTFC.js","assets/semver-D1BuALCK.js","assets/banners.store-DikUBA5o.js","assets/banners-UK5wWFEI.css","assets/assistant.store-BtplYeLr.js","assets/NodeIcon-BYTrCZAx.js","assets/NodeIcon-eaW3atZy.css","assets/ItemsRenderer-CejxMtdi.js","assets/useActions-BeWbvBcf.js","assets/ItemsRenderer-Cr6c6uYy.css","assets/chatPanel.store-BZOiAoXE.js","assets/CommunityNodeUpdateInfo-BbQHVpK8.js","assets/CommunityNodeUpdateInfo-kPDiipF5.css","assets/NodeCreator-JpUhf2LN.css"])))=>i.map(i=>d[i]);
import { D as createElementBlock, E as createCommentVNode, G as nextTick, Gt as unref, M as createVNode, N as defineAsyncComponent, P as defineComponent, Sn as toDisplayString, T as createBlock, _ as Fragment, _n as normalizeClass, bt as withCtx, et as openBlock, j as createTextVNode, w as createBaseVNode, y as Suspense } from "./vue.runtime.esm-bundler-DDuXT-9r.js";
import { gt as useI18n } from "./_MapCache-Dw6Llo22.js";
import { Ln as N8nButton_default, Tt as N8nIconButton_default, st as AskAssistantIcon_default, vt as N8nTooltip_default } from "./src-DsaRShTK.js";
import "./en-BYTsM8fR.js";
import { t as __vitePreload } from "./preload-helper-CR0ecmWK.js";
import { t as __plugin_vue_export_helper_default } from "./_plugin-vue_export-helper-BwBpWJRZ.js";
import "./truncate-7XypoloM.js";
import "./icon-DmIKo5zZ.js";
import "./overlay-C02_kK8Y.js";
import "./empty-RTEMfF4N.js";
import "./useMessage-hiDKbQYH.js";
import { Ni as useUIStore, t as useTelemetry, v as getMidCanvasPosition } from "./useTelemetry-DedEakwK.js";
import "./useToast-LQk5sJH0.js";
import "./sanitize-html-Cfxibw1n.js";
import "./path-browserify-CTmc1OxV.js";
import { br as DEFAULT_STICKY_HEIGHT, io as NODE_CREATOR_OPEN_SOURCES, xn as STICKY_NODE_TYPE, xr as DEFAULT_STICKY_WIDTH } from "./constants-C5Tj3dXQ.js";
import "./merge-Dnkh07HW.js";
import { i as useBuilderStore, t as useAssistantStore } from "./assistant.store-BtplYeLr.js";
import "./dateformat-D7TIhVd4.js";
import "./useDebounce-DFa0sFlC.js";
import "./useExternalHooks-DSSTcvme.js";
import { t as useChatPanelStore } from "./chatPanel.store-BZOiAoXE.js";
import "./npsSurvey.store-ylMA9JZD.js";
import "./cloudPlan.store-DT2PFV79.js";
import "./templates.store-Bt8tpLPx.js";
import { t as useFocusPanelStore } from "./focusPanel.store-CPU1uVq-.js";
import "./useWorkflowSaving-GL9jVODA.js";
import "./retry-sDkwzrPY.js";
import "./executions.store-D2xBmTFC.js";
import "./useRunWorkflow-BLqckUNU.js";
import "./usePinnedData-BsuIBj2b.js";
import "./nodeCreator.store-C1GkQ_2N.js";
import "./nodeIcon-ChI78z3o.js";
import "./useClipboard-BkbYyMX3.js";
import "./useCanvasOperations-Cxgmc_7x.js";
import "./folders.store-CmYx4Y_F.js";
import "./NodeIcon-BYTrCZAx.js";
import { t as KeyboardShortcutTooltip_default } from "./KeyboardShortcutTooltip-qZ5YD2sD.js";
import "./versions.store-Bon8LCsf.js";
import "./usePageRedirectionHelper-C8YGBi-F.js";
import "./dataTable.store-BC5cIRU-.js";
import "./useBugReporting-B8EUAXTd.js";
import "./uniqBy-DAOmWpEt.js";
import "./ProjectIcon-CHk5fa0e.js";
import "./CredentialIcon-CwCVRpiA.js";
import "./FileSaver.min-DaytN8ae.js";
import "./sortByProperty-CUhVU03p.js";
import "./useRecentResources-D1-21efA.js";
import { t as useActions } from "./useActions-BeWbvBcf.js";
import { t as useCommandBar } from "./useCommandBar-DOChPPYZ.js";
import "./useWorkflowActivate-BSpG7MxI.js";
var NodeCreation_vue_vue_type_script_setup_true_lang_default = /* @__PURE__ */ defineComponent({
	__name: "NodeCreation",
	props: {
		nodeViewScale: {},
		createNodeActive: {
			type: Boolean,
			default: false
		},
		focusPanelActive: { type: Boolean }
	},
	emits: [
		"addNodes",
		"toggleNodeCreator",
		"close"
	],
	setup(__props, { emit: __emit }) {
		const LazyNodeCreator = defineAsyncComponent(async () => await __vitePreload(() => import("./NodeCreator-DEx3Dj6j.js"), __vite__mapDeps([0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31,32,33,34,35,36,37,38,39,40,41,42,43,44])));
		const props = __props;
		const emit = __emit;
		const uiStore = useUIStore();
		const focusPanelStore = useFocusPanelStore();
		const i18n = useI18n();
		const telemetry = useTelemetry();
		const assistantStore = useAssistantStore();
		const builderStore = useBuilderStore();
		const chatPanelStore = useChatPanelStore();
		const { isEnabled: isCommandBarEnabled } = useCommandBar();
		const { getAddedNodesAndConnections } = useActions();
		function openNodeCreator() {
			emit("toggleNodeCreator", {
				source: NODE_CREATOR_OPEN_SOURCES.ADD_NODE_BUTTON,
				createNodeActive: true
			});
		}
		function addStickyNote() {
			if (document.activeElement) document.activeElement.blur();
			const offset = [...uiStore.nodeViewOffsetPosition];
			const position = getMidCanvasPosition(props.nodeViewScale, offset);
			position[0] -= 240 / 2;
			position[1] -= 160 / 2;
			emit("addNodes", getAddedNodesAndConnections([{
				type: STICKY_NODE_TYPE,
				position
			}]));
		}
		function closeNodeCreator(hasAddedNodes = false) {
			if (props.createNodeActive) emit("toggleNodeCreator", {
				createNodeActive: false,
				hasAddedNodes
			});
			emit("close");
		}
		function nodeTypeSelected(value) {
			emit("addNodes", getAddedNodesAndConnections(value));
			closeNodeCreator(true);
		}
		function toggleFocusPanel() {
			focusPanelStore.toggleFocusPanel();
			telemetry.track(focusPanelStore.focusPanelActive ? "User opened focus panel" : "User closed focus panel", {
				source: "canvasButton",
				parameters: focusPanelStore.focusedNodeParametersInTelemetryFormat
			});
		}
		async function onAskAssistantButtonClick() {
			if (builderStore.isAIBuilderEnabled) await chatPanelStore.toggle({ mode: "builder" });
			else await chatPanelStore.toggle({ mode: "assistant" });
			if (chatPanelStore.isOpen) assistantStore.trackUserOpenedAssistant({
				source: "canvas",
				task: "placeholder",
				has_existing_session: !assistantStore.isSessionEnded
			});
		}
		function openCommandBar(event) {
			event.stopPropagation();
			nextTick(() => {
				const keyboardEvent = new KeyboardEvent("keydown", {
					key: "k",
					code: "KeyK",
					metaKey: true,
					bubbles: true,
					cancelable: true
				});
				document.dispatchEvent(keyboardEvent);
			});
		}
		return (_ctx, _cache) => {
			return openBlock(), createElementBlock(Fragment, null, [!_ctx.createNodeActive ? (openBlock(), createElementBlock("div", {
				key: 0,
				class: normalizeClass(_ctx.$style.nodeButtonsWrapper)
			}, [
				createVNode(KeyboardShortcutTooltip_default, {
					label: unref(i18n).baseText("nodeView.openNodesPanel"),
					shortcut: { keys: ["Tab"] },
					placement: "left"
				}, {
					default: withCtx(() => [createVNode(unref(N8nIconButton_default), {
						size: "large",
						icon: "plus",
						type: "tertiary",
						"data-test-id": "node-creator-plus-button",
						onClick: openNodeCreator
					})]),
					_: 1
				}, 8, ["label"]),
				unref(isCommandBarEnabled) ? (openBlock(), createBlock(KeyboardShortcutTooltip_default, {
					key: 0,
					label: unref(i18n).baseText("nodeView.openCommandBar"),
					shortcut: {
						keys: ["k"],
						metaKey: true
					},
					placement: "left"
				}, {
					default: withCtx(() => [createVNode(unref(N8nIconButton_default), {
						size: "large",
						icon: "search",
						type: "tertiary",
						"data-test-id": "command-bar-button",
						onClick: openCommandBar
					})]),
					_: 1
				}, 8, ["label"])) : createCommentVNode("", true),
				createVNode(KeyboardShortcutTooltip_default, {
					label: unref(i18n).baseText("nodeView.addStickyHint"),
					shortcut: {
						keys: ["s"],
						shiftKey: true
					},
					placement: "left"
				}, {
					default: withCtx(() => [createVNode(unref(N8nIconButton_default), {
						size: "large",
						type: "tertiary",
						icon: "sticky-note",
						"data-test-id": "add-sticky-button",
						onClick: addStickyNote
					})]),
					_: 1
				}, 8, ["label"]),
				createVNode(KeyboardShortcutTooltip_default, {
					label: unref(i18n).baseText("nodeView.openFocusPanel"),
					shortcut: {
						keys: ["f"],
						shiftKey: true
					},
					placement: "left"
				}, {
					default: withCtx(() => [createVNode(unref(N8nIconButton_default), {
						type: "tertiary",
						size: "large",
						icon: "panel-right",
						class: normalizeClass(_ctx.focusPanelActive ? _ctx.$style.activeButton : ""),
						active: _ctx.focusPanelActive,
						"data-test-id": "toggle-focus-panel-button",
						onClick: toggleFocusPanel
					}, null, 8, ["class", "active"])]),
					_: 1
				}, 8, ["label"]),
				unref(chatPanelStore).canShowAiButtonOnCanvas ? (openBlock(), createBlock(unref(N8nTooltip_default), {
					key: 1,
					placement: "left"
				}, {
					content: withCtx(() => [createTextVNode(toDisplayString(unref(i18n).baseText("aiAssistant.tooltip")), 1)]),
					default: withCtx(() => [createVNode(unref(N8nButton_default), {
						type: "tertiary",
						size: "large",
						square: "",
						class: normalizeClass(_ctx.$style.icon),
						"data-test-id": "ask-assistant-canvas-action-button",
						onClick: onAskAssistantButtonClick
					}, {
						default: withCtx(() => [createBaseVNode("div", null, [createVNode(unref(AskAssistantIcon_default), { size: "large" })])]),
						_: 1
					}, 8, ["class"])]),
					_: 1
				})) : createCommentVNode("", true)
			], 2)) : createCommentVNode("", true), (openBlock(), createBlock(Suspense, null, {
				default: withCtx(() => [createVNode(unref(LazyNodeCreator), {
					active: _ctx.createNodeActive,
					onNodeTypeSelected: nodeTypeSelected,
					onCloseNodeCreator: closeNodeCreator
				}, null, 8, ["active"])]),
				_: 1
			}))], 64);
		};
	}
});
var NodeCreation_vue_vue_type_style_index_0_lang_module_default = {
	nodeButtonsWrapper: "_nodeButtonsWrapper_13o6j_123",
	icon: "_icon_13o6j_134",
	activeButton: "_activeButton_13o6j_143"
};
var NodeCreation_default = /* @__PURE__ */ __plugin_vue_export_helper_default(NodeCreation_vue_vue_type_script_setup_true_lang_default, [["__cssModules", { "$style": NodeCreation_vue_vue_type_style_index_0_lang_module_default }]]);
export { NodeCreation_default as default };
