// @n8n/n8n-nodes-langchain/dist/nodes/vendors/OpenAi/OpenAi.node.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var OpenAi_node_exports = {};
__export(OpenAi_node_exports, {
  OpenAi: () => OpenAi
});
module.exports = __toCommonJS(OpenAi_node_exports);
var import_n8n_workflow = require("n8n-workflow");
var import_description = require("./helpers/description");
var import_OpenAiV1 = require("./v1/OpenAiV1.node");
var import_OpenAiV2 = require("./v2/OpenAiV2.node");
class OpenAi extends import_n8n_workflow.VersionedNodeType {
  constructor() {
    const baseDescription = {
      displayName: "OpenAI",
      name: "openAi",
      icon: { light: "file:openAi.svg", dark: "file:openAi.dark.svg" },
      group: ["transform"],
      defaultVersion: 2,
      subtitle: `={{(${import_description.prettifyOperation})($parameter.resource, $parameter.operation)}}`,
      description: "Message an assistant or GPT, analyze images, generate audio, etc.",
      codex: {
        alias: [
          "LangChain",
          "ChatGPT",
          "Sora",
          "DallE",
          "whisper",
          "audio",
          "transcribe",
          "tts",
          "assistant"
        ],
        categories: ["AI"],
        subcategories: {
          AI: ["Agents", "Miscellaneous", "Root Nodes"]
        },
        resources: {
          primaryDocumentation: [
            {
              url: "https://docs.n8n.io/integrations/builtin/app-nodes/n8n-nodes-langchain.openai/"
            }
          ]
        }
      }
    };
    const nodeVersions = {
      1: new import_OpenAiV1.OpenAiV1(baseDescription),
      1.1: new import_OpenAiV1.OpenAiV1(baseDescription),
      1.2: new import_OpenAiV1.OpenAiV1(baseDescription),
      1.3: new import_OpenAiV1.OpenAiV1(baseDescription),
      1.4: new import_OpenAiV1.OpenAiV1(baseDescription),
      1.5: new import_OpenAiV1.OpenAiV1(baseDescription),
      1.6: new import_OpenAiV1.OpenAiV1(baseDescription),
      1.7: new import_OpenAiV1.OpenAiV1(baseDescription),
      1.8: new import_OpenAiV1.OpenAiV1(baseDescription),
      2: new import_OpenAiV2.OpenAiV2(baseDescription)
    };
    super(nodeVersions, baseDescription);
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  OpenAi
});
